minimint-core/src/epoch.rs

```diff
@@ -0,0 +1,217 @@
+use crate::transaction::Transaction;
+use bitcoin_hashes::sha256::Hash as Sha256;
+use itertools::Itertools;
+use minimint_api::encoding::{Decodable, DecodeError, Encodable};
+use minimint_api::{BitcoinHash, FederationModule, PeerId};
+use minimint_derive::UnzipConsensus;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use threshold_crypto::{PublicKey, Signature, SignatureShare};
+
+#[derive(
+    Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize, UnzipConsensus, Encodable, Decodable,
+)]
+pub enum ConsensusItem {
+    EpochInfo(EpochSignatureShare),
+    Transaction(Transaction),
+    Mint(<minimint_mint::Mint as FederationModule>::ConsensusItem),
+    Wallet(<minimint_wallet::Wallet as FederationModule>::ConsensusItem),
+    LN(<minimint_ln::LightningModule as FederationModule>::ConsensusItem),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignatureShare(pub SignatureShare);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignature(pub Signature);
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct EpochHistory {
+    pub outcome: MerkleOutcome,
+    pub hash: Sha256,
+    pub signature: Option<EpochSignature>,
+    pub epoch: u64,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct MerkleOutcome {
+    pub last_hash: Option<Sha256>,
+    pub items: Vec<(PeerId, Vec<ConsensusItem>)>,
+}
+
+impl MerkleOutcome {
+    pub fn hash(&self) -> Sha256 {
+        let mut bytes: Vec<u8> = vec![];
+        self.consensus_encode(&mut bytes).unwrap();
+        Sha256::hash(&bytes)
+    }
```

Comment:
This doesn't seem right

Reply:
I do have some unit tests for the hashing, but is there something specifically that seems wrong?

Reply:
Oh nevermind. I thought it was just hashing an empty vec 😆 . I understand now.